sys/hoon.hoon

```diff
@@ -11598,58 +11348,3387 @@
+::  - `%option` -- a union of types which are all constant atoms.
+::  - `%union` -- a union of types which are all instances (cells whose
+::    head is a constant atom).
+::  - `%junction` -- a union of an atom type and a cell type.
+::  - `%conjunction` -- a union of two cell types, one of them %wide
+::     and the other %tall.
+::  - `%misjunction` -- any other union type. There's no efficient way
+::    to tell which branch to take when analyzing a fork which is a
+::    %misjunction, and the type is probably improperly constructed.
+::
++$  xrole
+  $@  $?  %void  %noun  %atom  %tall  %wide  ==
+  $%  [%constant =atom]
+      [%instance =atom]
+      [%option =(map atom xkey)]
+      [%union =(map atom xkey)]
+      [%junction flat=xkey deep=xkey]
+      [%conjunction wide=xkey tall=xkey]
+      [%misjunction one=xkey two=xkey]
+  ==
+::
+::  This is just a utility type, it encodes the "battery" structure
+::  within a core.
+::
+::  It's a map from chapter names to the documentation and arms within
+::  that chapter.
+::
++$  xbat  (map term (pair what (map term xkey)))
+::
+::  A recipe tells us how a type was constructed.
+::
+::  - `%direct` is a simple type like `term`, or `xray`.
+::  - `%synthetic` is a constructed type, like `(list @)`.
+::
++$  recipe
+  $%  [%direct =term]
+      [%synthetic =term =(list xkey)]
+  ==
+::
+::  A `xpat` is high-level information about the shape of a type. This
+::  is used for printing data.
+::
+::  This is fairly heuristic. [%a %b %c ~] is recognized as a `path`,
+::  `[3 ~[4 5 6]]` is recognized as a list, etc.
+::
+::  Most of the xpats have names that make their purpose obvious:
+::  for example, the %tape xpat means that data of type type can be
+::  printed as if it had the `tape` type. However, `%gear` and `%gate`
+::  might not be entirely obvious.
+::
+::  - The %gear xpat is any core with a cell subject.
+::  - The %gate xpat is a core that looks like a gate.
+::
++$  xpat
+  $@  ?(%hoon %manx %json %nock %path %plum %skin %spec %tape %tour %type %vase)
+  $%  [%gate sample=xkey product=xkey]
+      [%gear sample=xkey context=xkey batt=xbat]
+      [%list item=xkey]
+      [%tree item=xkey]
+      [%unit item=xkey]
+  ==
+::
+::  Left-fold over a list.
+::
+++  fold
+   |*  [state=mold elem=mold]
+   |=  [[st=state xs=(list elem)] f=$-([state elem] state)]
+   ^-  state
+   |-
+   ?~  xs  st
+   =.  st  (f st i.xs)
+   $(xs t.xs, st st)
+::
+::  This is basically a `mapM` over a list using the State monad.
+::
+::  Another way to think about this is that it is the same as `turn`,
+::  except that a state variable `st` is threaded through the
+::  execution. The list is processed from left to right.
+::
+++  traverse
+  |*  [state=mold in=mold out=mold]
+  |=  [[st=state xs=(list in)] f=$-([state in] [out state])]
+  ^-  [(list out) state]
+  ?~  xs  [~ st]
+  =^  r   st  (f st i.xs)
+  =^  rs  st  $(xs t.xs, st st)
+  [[r rs] st]
```

Comment:
Is this different from `+spin` (other than explicit types)?

Reply:
Nope. Added a comment.